File: apps/worker/src-go/internal/runtime/heartbeat.go
package runtime

import (
	"context"
	"log/slog"
	"sync/atomic"
	"time"

	"linky-worker/src-go/internal/job"
	redisclient "linky-worker/src-go/internal/redis"
)

func RunHeartbeat(ctx context.Context, rdb *redisclient.Client, workerID string, logger *slog.Logger, stopping *atomic.Bool) {
	if err := rdb.RefreshHeartbeat(ctx, workerID); err != nil {
		logger.Warn("heartbeat refresh failed", "workerId", workerID, "error", err)
	}

	ticker := time.NewTicker(job.WorkerHeartbeatRefresh)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			clearCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
			if err := rdb.ClearHeartbeat(clearCtx, workerID); err != nil {
				logger.Warn("heartbeat clear failed", "workerId", workerID, "error", err)
			}
			cancel()
			return
		case <-ticker.C:
			if stopping.Load() {
				continue
			}
			if err := rdb.RefreshHeartbeat(ctx, workerID); err != nil {
				logger.Warn("heartbeat refresh failed", "workerId", workerID, "error", err)
			}
		}
	}
}

File: apps/worker/src/queues/heartbeat.ts
import type { Logger } from "@ws/logger";
import {
  clearHeartbeat,
  refreshHeartbeat,
  type RedisListClient,
} from "@ws/sdk-internal";
import { WORKER_HEARTBEAT_REFRESH_MS } from "@ws/shared-types";

export type HeartbeatHandle = {
  stop: () => Promise<void>;
};

export function startHeartbeat(options: {
  client: RedisListClient;
  workerId: string;
  logger: Logger;
  intervalMs?: number;
}): HeartbeatHandle {
  const { client, workerId, logger } = options;
  const intervalMs = options.intervalMs ?? WORKER_HEARTBEAT_REFRESH_MS;

  let stopped = false;

  const tick = async (): Promise<void> => {
    try {
      await refreshHeartbeat(client, workerId);
    } catch (error: unknown) {
      const err = error instanceof Error ? error : new Error(String(error));
      logger.warn(err, "heartbeat refresh failed (workerId=%s)", workerId);
    }
  };

  void tick();
  const timer = setInterval(() => {
    void tick();
  }, intervalMs);

  return {
    async stop(): Promise<void> {
      if (stopped) return;
      stopped = true;
      clearInterval(timer);
      try {
        await clearHeartbeat(client, workerId);
      } catch (error: unknown) {
        const err = error instanceof Error ? error : new Error(String(error));
        logger.warn(err, "heartbeat clear failed (workerId=%s)", workerId);
      }
    },
  };
}
